### ISSUE TITLE
ProtocolException with iOS .NET 6.0 Application: The number of bytes available is inconsistent with the HTTP Content-Length header.

### ISSUE THREAD

**Issue by chamons** [dotnet/wcf#4903]

**Comment by danmoseley** [comment]
I think this belongs in dotnet/wcf

**Comment by DuncWatts** [comment]
Note this works fine in a .NET 6 WPF application, it's only the .NET 6 iOS build that has this issue so far. I've yet to try it on Android or MacOS.

**Comment by danmoseley** [comment]
cc @radical @steveisok since this may be iOS specific and therefore not a WCF issue - perhaps they have seen something similar.

**Comment by chamons** [comment]
Given the lack of any Xamarin iOS types in the stack is the reason I bounced it over to the runtime at first.

**Comment by chamons** [comment]
I just checked, and this works fine on macOS Desktop with NET6.

**Comment by chamons** [comment]
[This](https://github.com/dotnet/wcf/blob/main/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/HttpResponseMessageHelper.cs#L218-L228) appears to be the location of the exception.

**Comment by steveisok** [comment]
@simonrozsival Please give this a look and see if you can find anything.

**Comment by simonrozsival** [comment]
I managed to reproduce the issue locally. When the exception that @chamons pointed out is throw, the state is:
- `bytesRead == 0`
- `_contentLength == 348`
- `offset == 325`
- `count == 23`
- `buffer` contains 325 bytes corresponding to a UTF-8 encoded string `<?xml version="1.0" encoding="utf-8"?><soap:Envelope ...</soap:Envelope>`

The raw HTTP response is:

```yaml
HTTP/1.1 200 OK
Cache-Control: private, max-age=0
Content-Type: text/xml; charset=utf-8
Content-Encoding: gzip
Vary: Accept-Encoding
Server: Microsoft-IIS/10.0
X-AspNet-Version: 2.0.50727
X-Powered-By: ASP.NET
X-Powered-By-Plesk: PleskWin
Date: Thu, 08 Sep 2022 07:57:57 GMT
Content-Length: 348

<?xml version="1.0" encoding="utf-8"?><soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema"><soap:Body><AddResponse xmlns="http://tempuri.org/"><AddResult>3</AddResult></AddResponse></soap:Body></soap:Envelope>
```
As far as I can tell the actual size of the body is 325 bytes. I think the `ProtocolException` is the correct behavior in this case because the server responded with incorrect `Content-Length` (even though it is not consistent with legacy Xamarin which seems to ignore this mismatch). The code in WCF that checks the mismatch is  7 years old and the `ReadAsync` code in `NSUrlSessionHandler+NSUrlSessionDataTaskStream` hasn't changed in 4 years so this doesn't seem like a recent regression.

@chamons can you share the macOS desktop project so I can try to check why that behaves differently?

**Comment by DuncWatts** [comment]
@simonrozsival I've added a WPF .NET 6 application to my example that isn't throwing a ProtocolException if that's of any help.

[WcfExampleWithWPF.zip](https://github.com/dotnet/wcf/files/9525709/WcfExampleWithWPF.zip)

**Comment by DuncWatts** [comment]
I've run Fiddler against the WPF build and in this instance it's returning Content-Length: 325 so assume it must be something in how it's making the request.

Request headers:
```
POST /calculator.asmx HTTP/1.1
Host: www.dneonline.com
Cache-Control: no-cache, max-age=0
SOAPAction: "http://tempuri.org/Add"
Accept-Encoding: gzip, deflate
Content-Type: text/xml; charset=utf-8
Content-Length: 161
```

Response headers:
```
HTTP/1.1 200 OK
Cache-Control: private, max-age=0
Content-Type: text/xml; charset=utf-8
Vary: Accept-Encoding
Server: Microsoft-IIS/10.0
X-AspNet-Version: 2.0.50727
X-Powered-By: ASP.NET
X-Powered-By-Plesk: PleskWin
Date: Thu, 08 Sep 2022 09:50:39 GMT
Content-Length: 325
```

**Comment by simonrozsival** [comment]
@DuncWatts That's interesting. I ran the request a couple more times and I got three different `Content-Length` values for the same request (identical headers to yours) with a body length of 325 bytes: `262`, `325`, `348`. It seems there's something wrong with that server.

Anyway, I found a workaround that might help. If you can, add this property to your iOS project: `<UseNativeHttpHandler>false</UseNativeHttpHandler>`. If you depend on the behavior of `NSUrlSessionHandler` somewhere, you can will need to instantiate it yourself and pass it to  `HttpClient` explicitly through its constructor. This may not be suitable for all projects though.

The `SocketsHttpHandler`, that is used when the native one is disabled, behaves differently: the response's `content.Headers.ContentLength` doesn't have a value (even though the response contains it) so the `ReadChunkedBufferedMessageAsync` method is called instead of `ReadBufferedMessageAsync`. The `ReadChunkedBufferedMessageAsync` method ignores the `Content-Length` header and reads the whole response body.

https://github.com/dotnet/wcf/blob/acb1157835acda59bde51cb1e64f1340619379e7/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/HttpResponseMessageHelper.cs#L178-L183

I'm not sure if this is the desired behavior of `SocketsHttpHandler` or not, but it's definitely an inconsistency between the platforms. There were some improvements to the `Content-Length` behavior recently (for example https://github.com/dotnet/runtime/pull/62541) so I'll try it with the latest preview .NET 7 later as well.

**Comment by simonrozsival** [comment]
@karelz is the behavior of `SocketsHttpHandler` and the `HttpResponseMessage` it returns described above the expected behavior or is that a bug?

**Comment by mconnew** [comment]
WCF has the ability to enable modifying the HttpMessageHandler stack. There's a [blog post here](https://justsimplycode.com/2019/11/02/disable-header-100-continue-in-net-core-wcf-client/) that demonstrates how to modify the behavior of the HttpClientHandler where you can wrap it if needed. If you need to keep using the native implementation, you could wrap it with an implementation which removes the Content-Length header once the response has been received before returning the response message to WCF. This will cause WCF to switch to a mode where it keeps reading the content until it either reaches the end, or it hits the configured MaxReceivedMessageSize byte limit.

**Comment by mconnew** [comment]
@simonrozsival, @DuncWatts, @karelz, which repo is the best place to track this? I think it's been sufficiently investigated to know this is not an issue WCF's implementation or usage of HttpClient.

**Comment by simonrozsival** [comment]
If @karelz confirms this is a bug then we should move it to dotnet/runtime.

**Comment by karelz** [comment]
@simonrozsival can you please restate what exactly is the `SocketsHttpHandler` behavior that you suspect may be a bug? (I tried to gather it from previous comment, but it was not 100% clear to me sadly :() Thanks!

If we find a bug in some component, I would recommend to create a new bug with specific details instead of this one which is IMO not super-clear at the moment. (there is of course always an option it is just me being slow ;))

**Comment by simonrozsival** [comment]
@karelz when `HttpClient` is used with `SocketsHttpHandler`, the response message's `content.Headers.ContentLength` doesn't have a value even though the HTTP response contained the `Content-Length` header. This doesn't happen when `NSUrlSessionHandler` is used instead. I think I'll just go ahead and create a new issue in dotnet/runtime with the details.

**Comment by DuncWatts** [comment]
Removing the Content-Length header using a DelegatingHandler works a charm thank you.  The previous workaround of setting UseNativeHttpHandler to false did work but triggered a lot of errors with the linker which I was struggling to resolve.

**Comment by simonrozsival** [comment]
tl;dr: The root of the problem is gzip compression of the response. The server works correctly. It's not a bug in .NET. It's possibly a bug in WCF and/or `NSUrlSessionHandler`.

@karelz I won't open any issue in dotnet/runtime after all. Sorry for the confusion 😄 

---

I did some more digging and I realized that my previous conclusions were incorrect. I overlooked the fact that the response is gzipped and so the Content-Length contains the size of the compressed payload (interesting in this case the gzipped payload is 348 B instead of the original 325 B). WCF sets the `AutomaticDecompression` property of the handler and it leaves all the decompression implementation details to the handler.

The .NET implementation (``) makes the `content.Headers.ContentLength` value inaccessible, while the `NSUrlSessionHandler` implementation returns the size of the compressed payload. WCF assumes that when the `ContentLength` value is available, it's the size of the plaintext content and not necessarily the exact value of the `Content-Length` header value.

@mconnew The docs of the [`ContentLength`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpcontentheaders.contentlength?view=net-6.0#system-net-http-headers-httpcontentheaders-contentlength) property say that it should contain the value of `Content-Length` and the Mac/iOS implementation sticks to the exact wording of the docs. I think WCF should ignore the value if the payload was compressed instead of relying on .NET's internal implementation. Or is there some documentation that would suggest it's a bug in the Mac/iOS implementation.

**Comment by MihaZupan** [comment]
`wcf` shouldn't be validating `Content-Length` here, it's either redundant or just wrong.

If decompression happened, or the server returned both the content length and chunked transfer encoding, or some handler somewhere changed the content/headers, this verification logic will be wrong.

`SocketsHttpHandler` will enforce the correctness of the response content length (it has to - it's the core of the protocol).
If the server sent us too many bytes, you won't see them when reading the content stream.
If the server sent us too few bytes, the `ReadAsync` on the content stream will throw.

**Comment by MihaZupan** [comment]
Ah, I see you just commented that decompression was indeed the issue.
This should be fixed in wcf. You should let the underlying handler handle the protocol correctness.
WCF shouldn't be enforcing the value of the `Content-Length` as it *can* be different from the number of bytes read from the content.

**Comment by mconnew** [comment]
What was the behavior with HttpWebRequest? We are using the exact same logic on .NET Framework.

**Comment by psudodev** [comment]
I am facing similar issue on Maui Android on .net 7 @chamons did you find any work around for this

**Comment by mconnew** [comment]
@rolfbjarne, the details are discussed in the [linked issue here.](https://github.com/dotnet/runtime/issues/80935) Basically NetFx, SocketsHttpHandler, CurlHandler, WinHttpHandler and now the android wrapper all remove the Content-Length header and the Content-Encoding header when using automatic decompression. This makes sense when you consider that when it automatically decompresses the content for you, the length doesn't tell you anything useful, and the encoding is irrelevant. Looks like iOS has different behavior. You can create the same behavior by using an extensibility point in WCF. If you [see this blog post](https://justsimplycode.com/2019/11/02/disable-header-100-continue-in-net-core-wcf-client/), it explains how to modify a request before it's sent. Basically copy the implementation but instead of modifying the Expect header before sending, remove the content length and encoding headers after receiving the response. These will be on reponse.Content.Headers as HttpClient places content related headers on the HttpContent object.
